tanner/reporting/hpfeeds.py

```diff
@@ -0,0 +1,224 @@
+import logging
+import struct
+import hashlib
+import sys
+import socket
+import time
+
+logger = logging.getLogger('pyhpfeeds')
+
+BUFSIZ = 16384
+
+OP_ERROR        = 0
+OP_INFO         = 1
+OP_AUTH         = 2
+OP_PUBLISH      = 3
+OP_SUBSCRIBE    = 4
+
+MAXBUF = 1024**2
+SIZES = {
+    OP_ERROR: 5+MAXBUF,
+    OP_INFO: 5+256+20,
+    OP_AUTH: 5+256+20,
+    OP_PUBLISH: 5+MAXBUF,
+    OP_SUBSCRIBE: 5+256*2,
+}
+
+__all__ = ["new", "FeedException"]
+
+
+class BadClient(Exception):
+    pass
+
+class FeedException(Exception):
+    pass
+
+class Disconnect(Exception):
+    pass
+
+# packs a string with 1 byte length field
+def strpack8(x):
+    if isinstance(x, str):
+        x = x.encode('latin1')
+    return struct.pack('!B', len(x)%0xff) + x
+
+
+# unpacks a string with 1 byte length field
+def strunpack8(x):
+    l = x[0]
+    return x[1:1+l], x[1+l:]
+
+
+def msghdr(op, data):
+    return struct.pack('!iB', 5+len(data), op) + data
+
+
+def msgpublish(ident, chan, data):
+    return msghdr(OP_PUBLISH, strpack8(ident) + strpack8(chan) + data.encode('latin1'))
+
+
+def msgsubscribe(ident, chan):
+    if isinstance(chan, str):
+        chan = chan.encode('latin1')
+    return msghdr(OP_SUBSCRIBE, strpack8(ident) + chan)
+
+
+def msgauth(rand, ident, secret):
+    auth_hash = hashlib.sha1(rand+secret.encode('latin1')).digest()
+    return msghdr(OP_AUTH, strpack8(ident) + auth_hash)
+
+
+class FeedUnpack(object):
+    def __init__(self):
+        self.buf = bytearray()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        return self.unpack()
+
+    def feed(self, data):
+        self.buf.extend(data)
+
+    def unpack(self):
+        if len(self.buf) < 5:
+            raise StopIteration('No message.')
+
+        ml, opcode = struct.unpack('!iB', self.buf[:5])
+        if ml > SIZES.get(opcode, MAXBUF):
+            raise BadClient('Not respecting MAXBUF.')
+
+        if len(self.buf) < ml:
+            raise StopIteration('No message.')
+
+        data = self.buf[5:ml]
+        del self.buf[:ml]
+        return opcode, data
+
+
+class HPC(object):
+    def __init__(self, host, port, ident, secret, timeout=3, reconnect=False, sleepwait=20):
+        self.host, self.port = host, port
+        self.ident, self.secret = ident, secret
+        self.timeout = timeout
+        self.reconnect = reconnect
+        self.sleepwait = sleepwait
+        self.brokername = 'unknown'
+        self.connected = False
+        self.stopped = False
+        self.s = None
+        self.unpacker = FeedUnpack()
+
+        self.tryconnect()
+
+    def send(self, data):
+        try:
+            self.s.sendall(data)
+        except socket.timeout:
+            logger.warn("Timeout while sending - disconnect.")
+            raise Disconnect()
+        except socket.error as e:
+            logger.warn("Socket error: %s", e)
+            raise Disconnect()
+
+    def tryconnect(self):
+        if not self.connected:
+            while True:
```

Comment:
I don't think `while True` is a good decision here. In case when the connection is not alive for a long time, tanner can't work because it's totally blocked. There are two ways - exit the tanner with logs and messages or mark the hpfeeds connection unsuccessful, log it and don't use later. We can use first way for now, it will be simpler.

I really appreciate your help and I completely want to merge your pr in master :)

Reply:
I know the `while True` is not a best idea and it can cause the problems, but it is from the original hpfeeds lib (https://github.com/rep/hpfeeds/blob/master/lib/hpfeeds.py). I wanted to discuss it with you but I was quite busy last days. I will try to improve it but right now I need to do some other work tasks. Sorry. I will make the fix ASAP.  
Thank you.